[PGNCCL] Use non-blocking mode by default in eager init

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138527
* #138860

### Why use non-blocking mode in eager init?
For overlapping comm init and model init, etc.
![image](https://github.com/user-attachments/assets/9b0bf7a9-be26-4d16-827b-dbe861f083cd)

### Why can we set non-blocking as default?
If the setting is dangling -- i.e. not passed in by user nor set via env -- `ProcessGroupNCCL` can have some preferred logic. And torch-level API semantics does not change whether the NCCL comm is blocking or non-blocking (handled within `ProcessGroupNCCL`).

### Why not make non-blocking default for lazy mode as well?
PR https://github.com/pytorch/pytorch/pull/137544 tried it.
Two reasons why that's not preferred today:
1. It is hard -- too big a blast.
2. There is no gain by doing lazy init in non-blocking mode, because the right next CPU call is a collective, and we will block there waiting for comm to be ready, so same effect as blocked init, no "opening" compared to eager mode.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[PGNCCL] Use non-blocking mode by default in eager init

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1059,6 +1058,41 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
+  // 1st priority: Respect the user's setting
+  if (options_->config.blocking != NCCL_CONFIG_UNDEF_INT) {
+    useNonblocking_ = options_->config.blocking == 0;
+    goto print_and_return;
```
goto! thats something i haven't seen in a while...

shuqiangzhang replied:
Almost never see goto in c++. Could use 'else if'? Dijkstra's publication of why [Go-to statement considered harmful](http://www.cs.utexas.edu/users/EWD/ewd02xx/EWD215.PDF) .

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1059,6 +1058,41 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
+  }
+  // 2nd priority: Respect the environment variable
+  if (nbEnv.has_value()) {
+    useNonblocking_ = nbEnv.value();
```
hm, do we need to check if bound device id exists here too? i thought we could not enable nonblocking without a bound id? maybe i just got confused by the bundling.

kwen2501 replied:
The env of non-blocking existed before we introduce bounded device.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1059,6 +1058,41 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
+    useNonblocking_ = true;
+    goto print_and_return;
+  }
+  // 4th priority: otherwise, nonblocking = false to preserve old behavior
```
lgtm, but i wonder if longer term, is this a path we'd want to add a 'deprecation warning' to start encouraging users to add device_id?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1059,6 +1058,41 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
+  useNonblocking_ = false;
+
+print_and_return:
+  LOG(INFO) << logPrefix()
```
nit: this is ok, but i think its also useful to log which of the config pathways took precedent, in case users want to debug why their nonblocking setting wasn't as expected

wconstab reviewed (approved):
overall lgtm.

shuqiangzhang commented:
Nit: Better to add a test case with eager init but without setting the ENV?

kwen2501 commented:
@shuqiangzhang Yeah, that's a good idea. I added a PR on top to parameterize tests in `test_c10d_nccl.py` to do both lazy and eager inits.

kwen2501 commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Seems to have introduce regressions on main, pull / linux-focal-cuda11.8-py3.10-gcc9 / test (distributed, 2, 3, linux.g4dn.12xlarge.nvidia.gpu) checking if revert will do" -c nosignal

kwen2501 committed:
Update on "[PGNCCL] Use non-blocking mode by default in eager init"


### Why use non-blocking mode in eager init?
For overlapping comm init and model init, etc.
![image](https://github.com/user-attachments/assets/9b0bf7a9-be26-4d16-827b-dbe861f083cd)

### Why can we set non-blocking as default?
If the setting is dangling -- i.e. not passed in by user nor set via env -- `ProcessGroupNCCL` can have some preferred logic. And torch-level API semantics does not change whether the NCCL comm is blocking or non-blocking (handled within `ProcessGroupNCCL`).

### Why not make non-blocking default for lazy mode as well?
PR https://github.com/pytorch/pytorch/pull/137544 tried it.
Two reasons why that's not preferred today:
1. It is hard -- too big a blast.
2. There is no gain by doing lazy init in non-blocking mode, because the right next CPU call is a collective, and we will block there waiting for comm to be ready, so same effect as blocked init, no "opening" compared to eager mode.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Use non-blocking mode by default in eager init"


### Why use non-blocking mode in eager init?
For overlapping comm init and model init, etc.
![image](https://github.com/user-attachments/assets/9b0bf7a9-be26-4d16-827b-dbe861f083cd)

### Why can we set non-blocking as default?
If the setting is dangling -- i.e. not passed in by user nor set via env -- `ProcessGroupNCCL` can have some preferred logic. And torch-level API semantics does not change whether the NCCL comm is blocking or non-blocking (handled within `ProcessGroupNCCL`).

### Why not make non-blocking default for lazy mode as well?
PR https://github.com/pytorch/pytorch/pull/137544 tried it.
Two reasons why that's not preferred today:
1. It is hard -- too big a blast.
2. There is no gain by doing lazy init in non-blocking mode, because the right next CPU call is a collective, and we will block there waiting for comm to be ready, so same effect as blocked init, no "opening" compared to eager mode.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Use non-blocking mode by default in eager init"


### Why use non-blocking mode in eager init?
For overlapping comm init and model init, etc.
![image](https://github.com/user-attachments/assets/9b0bf7a9-be26-4d16-827b-dbe861f083cd)

### Why can we set non-blocking as default?
If the setting is dangling -- i.e. not passed in by user nor set via env -- `ProcessGroupNCCL` can have some preferred logic. And torch-level API semantics does not change whether the NCCL comm is blocking or non-blocking (handled within `ProcessGroupNCCL`).

### Why not make non-blocking default for lazy mode as well?
PR https://github.com/pytorch/pytorch/pull/137544 tried it.
Two reasons why that's not preferred today:
1. It is hard -- too big a blast.
2. There is no gain by doing lazy init in non-blocking mode, because the right next CPU call is a collective, and we will block there waiting for comm to be ready, so same effect as blocked init, no "opening" compared to eager mode.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented:
@kwen2501 @wconstab I want to point out a fundamental issue due to the constraint of P2P/split. Currently P2P (creating new comms for P2P) does not work with ncclCommSplit because of its collective requirement (need to call no color split) on all ranks. That's another reason we need to separate the concept of eager init and split. If we bind eager/init to nonblocking together, it also means nonblocking mode cannot be used for p2p in current format

kwen2501 commented:
> @kwen2501 @wconstab I want to point out a fundamental issue due to the constraint of P2P/split. Currently P2P (creating new comms for P2P) does not work with ncclCommSplit because of its collective requirement (need to call no color split) on all ranks. That's another reason we need to separate the concept of eager init and split. If we bind eager/split to nonblocking together, it also means nonblocking mode cannot be used for p2p in current format

Discussed offline. There is a split-vs-P2P bug that exists before this PR. We shall fix that bug separately. @shuqiangzhang will file an issue.

Concretely, the bug is here:
https://github.com/pytorch/pytorch/blob/c6bb9b53f4d49dbfb2a19553731f8fb01e19dc96/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2361

When P2P comm creation invokes `getNCCLComm`, it may see a `split_from` options which is meant for the previous PG creation. Then the P2P comm creation may use `ncclCommSplit` and hang, because not all ranks join this call. The bug slips previously/today because there is no CI test with the following recipe: eager init + new group + P2P in that new group.

kwen2501 committed:
Update on "[PGNCCL] Use non-blocking mode by default in eager init"


### Why use non-blocking mode in eager init?
For overlapping comm init and model init, etc.
![image](https://github.com/user-attachments/assets/9b0bf7a9-be26-4d16-827b-dbe861f083cd)

### Why can we set non-blocking as default?
If the setting is dangling -- i.e. not passed in by user nor set via env -- `ProcessGroupNCCL` can have some preferred logic. And torch-level API semantics does not change whether the NCCL comm is blocking or non-blocking (handled within `ProcessGroupNCCL`).

### Why not make non-blocking default for lazy mode as well?
PR https://github.com/pytorch/pytorch/pull/137544 tried it.
Two reasons why that's not preferred today:
1. It is hard -- too big a blast.
2. There is no gain by doing lazy init in non-blocking mode, because the right next CPU call is a collective, and we will block there waiting for comm to be ready, so same effect as blocked init, no "opening" compared to eager mode.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing on ROCm' -c weird

distributed/test_c10d_nccl.py::ProcessGroupNCCLGroupTest::test_extra_cuda_context [GH job link](https://github.com/pytorch/pytorch/actions/runs/11529791397/job/32099048680) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/07e30eae2a8241e531890b6c9a33ab5a80c5ccaf)

kwen2501 commented:
@huydhn Thanks for the notification. It seems `test_extra_cuda_context` was flaky for rocm since 10/23, before my PR's landed. So I disabled that test. And relanding this PR.

kwen2501 commented:
@pytorchbot merge -f "test_extra_cuda_context was flaky for rocm since 10/23, before my PR's landed. So I disabled that test"
